File: chess-logic-api/src/models/King.ts
import { PieceColor, PieceType } from "../Utilities/enums";
import Board from "./Board";
import Square from "./Square";
import Piece from "./Piece";


export default class King extends Piece {
  private pieceType = PieceType.King;

  constructor(color: PieceColor, currentSquare: Square) {
    super(color, currentSquare);
  }

  public getValidMoves(board: Board): Square[] {
    const validMoves: Square[] = [];
    
    return validMoves;
  }
}

File: chess-logic-api/src/wasm/cpp/models/King.cpp
#include "King.h"
#include "../utils/structs/Move.h"

King::King(PieceColor pieceColor, Square& currentSquare) : Piece(pieceColor, currentSquare) {
  pieceType = PieceType::King;
}

bool King::isValidMove(Board& board, Square& targetSquare) {
  if (!Piece::isValidMove(board, targetSquare)) {
    return false;
  }

  Move moves[] = {
    {1, 0},    // Up
    {1, 1},    // Up-Right
    {0, 1},    // Right
    {-1, 1},   // Down-Right
    {-1, 0},   // Down
    {-1, -1},  // Down-Left
    {0, -1},   // Left
    {1, -1}    // Up-Left
  };

  for (const auto& move : moves) {
    if (targetSquare.getRowIndex() == currentSquare.getRowIndex() + move.row &&
        targetSquare.getColIndex() == currentSquare.getColIndex() + move.col) {
      return true;
    }
  }

  return false;
}
